core/src/main/scala/com/databricks/labs/remorph/parsers/VisitorCoordinator.scala

```diff
@@ -0,0 +1,30 @@
+package com.databricks.labs.remorph.parsers
+import com.databricks.labs.remorph.parsers.tsql.DataTypeBuilder
+import org.antlr.v4.runtime.tree.ParseTreeVisitor
+
+/**
+ * <p>
+ *   An implementation of this class should provide an instance of each of the ParseTreeVisitors
+ *   required to build IR and each visitor should be initialized with a reference to the implementation
+ *   of this class so that the visitors can call each other without needing to know the specific implementation
+ *   of any visitor or creating circular dependencies between them.
+ * </p>
+ * <p>
+ *   Implementations can also supply other shared resources, builders, etc. via the same mechanism
+ * </p>
+ */
+abstract class VisitorCoordinator {
```

Comment:
keep it in the tsql package, for now - it doesn't make sense to introduce an abstraction in common package before it's mature enough and adopted by both parsers

Reply:
Fair point - I thought about that too. It is fairly clear that TSQL and Snowflake are following the same builder set and it seems that new dialects should follow the same pattern but neither are completely finished with IR building. So I will leave the abstract class but not apply the same to Snowflake at the moment. I can remove the abstract class if you prefer.